feat: support adding ResourceManagerTags to GCP resources

salasberryfin opened the pull request:
**What type of PR is this?**

kind/feature

**What this PR does / why we need it**

This PR allows users to bind tags to GCP resources, including Compute Instances, GKE clusters and attached disks. As described in the original issue #992, Tags are used for defining IAM policy conditions, Organization conditionals policies and integrating with Cloud billing for cost management, which are not supported by labels.

**Which issue(s) this PR fixes**

Fixes #992 

**Special notes for your reviewer**

With this solution, users are able to bind resource-manager tags to infrastructure resources. The Tag Key/Tag Value must already exist in GCP and CAPG converts the provided list of key/value (` resourceManagerTags`) into the unique IDs generated by GCP. For any tags that cannot be retrieved (whether it is a permission issue or the tags don't exist), an empty value will be returned, effectively ignoring the key/value pair. Any other valid tags will be added. 

This is an example of how a simple tag could be added to the yaml definition:
```
resourceManagerTags:
  - {parentID: "parent-project", key: "foo", value: "bar"}
```

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Users can now pass existing `resourceManagerTags` to bind them to cloud resources. If this is not provided, the behavior stays the same and no tags are added to the resource.
```

richardcase commented:
/ok-to-test

richardcase commented:
/test pull-cluster-api-provider-gcp-verify

salasberryfin commented:
I addressed some of the comments but my changes broke something. I'll be fixing it and updating the code as soon as possible.

salasberryfin commented:
Hopefully I addressed all comments and CI issues with the latest commit 🙂

richardcase commented on api/v1beta1/gcpcluster_types.go:
```diff
@@ -54,6 +54,14 @@ type GCPClusterSpec struct {
+	// +listType=map
+	// +listMapKey=key
+	// +optional
+	ResourceManagerTags []ResourceManagerTag `json:"resourceManagerTags,omitempty"`
```
I see `ResourceManagerTags` on the GCPMachine being used. However, i can't see where the `ResourceManagerTags` on GCPCluster/GCPManagedCluster are being used?

There is a pattern in other providers for tags where there is precedence. Tags can be supplied on the infra cluster (i.e. GCPCluster or GCPManagedCluster for GCP) or on the infra machine (i.e. GCPMachine). If they are both supplied then the tags on GCPMachine are used.

richardcase commented on cloud/scope/machine.go:
```diff
@@ -338,6 +338,9 @@ func (m *MachineScope) InstanceSpec(log logr.Logger) *compute.Instance {
 				m.ClusterGetter.Name(),
 			),
 		},
+		Params: &compute.InstanceParams{
```
This will tag the compute instances.

We should probably tag other GCP resources we are creating, like the actual GKE cluster:
https://cloud.google.com/kubernetes-engine/docs/how-to/tags

So we probably need to call [CreateTagBinding](https://cloud.google.com/go/docs/reference/cloud.google.com/go/resourcemanager/latest/apiv3#cloud_google_com_go_resourcemanager_apiv3_TagBindingsClient_CreateTagBinding) for other resources. wdyt?

bharath-b-rh replied:
Disks attached to Instance also supports Tags, I think we should include tags in disk params at [[1]](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/c0eb2c2901611d75e1adabf854fbb75353bf3ebf/cloud/scope/machine.go#L241) and [[2]](https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/c0eb2c2901611d75e1adabf854fbb75353bf3ebf/cloud/scope/machine.go#L255)?

richardcase commented on cloud/scope/machine.go:
```diff
@@ -338,6 +338,9 @@ func (m *MachineScope) InstanceSpec(log logr.Logger) *compute.Instance {
 			),
 		},
+		Params: &compute.InstanceParams{
+			ResourceManagerTags: infrav1.AddResourceManagerTags(context.TODO(), m.GCPMachine.Spec.ResourceManagerTags),
```
We should probably capture the additional permissions required in the docs somewhere:

- roles/resourcemanager.tagAdmin
- roles/resourcemanager.tagUser

bharath-b-rh replied:
I think, TagUser role should be enough since listing tags and creating tag bindings are the operations implementation is dependent on.

richardcase reviewed (commented):
Great work on this @salasberryfin .

Lets find some time to chat about this

bharath-b-rh commented on api/v1beta1/gcpmachine_types.go:
```diff
@@ -197,6 +197,14 @@ type GCPMachineSpec struct {
 	// +optional
 	AdditionalNetworkTags []string `json:"additionalNetworkTags,omitempty"`
 
+	// ResourceManagerTags is an optional set of tags to apply to GCP resources managed
```
Apologies for bring this up again.
I feel, lower camel case should be used here, the way corresponding field names appear in CRD and `kubectl explain`.

salasberryfin replied:
In favor of using upper case letters, it is what is used in similar fields in the struct, like `AdditionalNetworkTags` or `AdditionalLabels`.

bharath-b-rh commented on cloud/scope/machine.go:
```diff
@@ -338,6 +338,9 @@ func (m *MachineScope) InstanceSpec(log logr.Logger) *compute.Instance {
 			),
 		},
+		Params: &compute.InstanceParams{
+			ResourceManagerTags: infrav1.AddResourceManagerTags(context.TODO(), m.GCPMachine.Spec.ResourceManagerTags),
```
I think, we should have the merge functionality here for the tags defined in MachineTemplate.Spec and Machine.Spec, with those in Machine.Spec taking precedence.

salasberryfin commented:
Thanks for all the comments @richardcase @bharath-b-rh. I'm now working on addressing all the points, starting with aligning CAPG with what is done in CAPA in terms of tags (considering that resource manager tags have this added requirement of having to convert tag keys/values to API-valid identifiers). This will probably take me some time because there are a couple of things in my original changes I need to rethink. I hope I can have this updated as soon as possible. 🙂

salasberryfin commented:
Pushed a partial refactor of the initial proposal based on the suggestions. Some of the points I focused on:
- Tags from `GCPCluster`, `GCPManagedCluster` and `GCPMachine` are now considered.
- Attached disks are also tagged.
- A tag binding is created for the cluster itself.

I still need to test a few scenarios, including the tag binding on GKE clusters, and fix some CI issues, but hopefully the bulk of the work is ready.

salasberryfin commented:
While waiting for CI to pass (I'll fix any errors if they occur), I consider this ready to be reviewed. It now supports adding tags to GKE clusters as well. Google's [`resourcemanagerpb.TagBinding` on pkg.go.dev](https://pkg.go.dev/cloud.google.com/go/resourcemanager@v1.9.1/apiv3/resourcemanagerpb#TagBinding) states:
```
A TagBinding represents a connection between a TagValue and a cloud resource Once a TagBinding is created, the TagValue is applied to all the descendants of the Google Cloud resource.
```
but Compute Instances that form the cluster don't seem to get the tag propagated. Any ideas or previous experience on why this happens? I expect these instances to be part of the descendants to which tags are propagated.

bharath-b-rh commented:
> Compute Instances that form the cluster don't seem to get the tag propagated. Any ideas or previous experience on why this happens? I expect these instances to be part of the descendants to which tags are propagated.

If I've understood the query and based on my understanding on GCP Tags, Tags are propagated to the resources in the Google Cloud Hierarchy, if the Organization/Folder/Project in which Compute Instances are created had any tags attached, it would be inherited by the instances. But here compute instance is part of the cluster and is not descendant/child of the cluster, so tags are not inherited.

salasberryfin commented:
Can someone take a look at this updated version of the PR? Hopefully it adds the required functionality.

richardcase commented:
This is only adding optional API fields so:

/override pull-cluster-api-provider-gcp-apidiff

richardcase commented on api/v1beta1/tags.go:
```diff
@@ -0,0 +1,157 @@
+
+// Bind creates a TagBinding between a TagValue and a Google Cloud resource.
+// If any of the SDK calls fail, the error is logged and no action is taken.
+func (t ResourceManagerTags) Bind(ctx context.Context, location, name string) {
```
We probably want to move the bind to a "service" and outside the API definition.

richardcase replied:
https://github.com/kubernetes-sigs/cluster-api-provider-gcp/tree/8bb4d822317484d2f0e1cafc4fee6bf17376ed80/cloud/services

richardcase replied:
(or into the existing services)

richardcase commented on api/v1beta1/tags.go:
```diff
@@ -0,0 +1,157 @@
+func (t ResourceManagerTags) Bind(ctx context.Context, location, name string) {
+	log := log.FromContext(ctx)
+	endpoint := fmt.Sprintf("%s-cloudresourcemanager.googleapis.com:443", location)
+	c, err := resourcemanager.NewTagBindingsClient(
```
When creating the client for the resource manager we should do something similar to how we create other clients:

https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/8bb4d822317484d2f0e1cafc4fee6bf17376ed80/cloud/scope/clients.go#L91

This will ensure that "per cluster" credentials are handled.

richardcase commented on api/v1beta1/tags.go:
```diff
@@ -0,0 +1,157 @@
+	for _, tag := range t {
+		tagValue, err := getTagValues(ctx, tag)
+		if err != nil {
+			log.Error(err, "failed to retrieve tag value")
```
Should we be returning an error instead and then let this be reported by the reconciler?

richardcase replied:
(applies to similar cases below)

richardcase commented on api/v1beta1/tags.go:
```diff
@@ -0,0 +1,157 @@
+// Convert converts the passed resource-manager tags to a GCP API valid format.
+// Tag keys and Tag Values will be created by the user and only the Tag bindings to the Compute Instance will be
+// handled by CAPG. If the Tag Key/Tag Value cannot be retrieved or no tags are provided, this will be empty and no tags will be added.
+func (t ResourceManagerTags) Convert(ctx context.Context) ResourceManagerTagsMap {
```
Maybe also move this to a "tags" service

salasberryfin commented:
`NewTagBindingsClient` is now created in `cloud/scope/clients.go` aligned with other client creation. In this case, due to the regional nature of some resource manager tagging operations, the endpoint (including location) needs to be appended to the default client options.

richardcase commented on cloud/services/shared/tags.go:
```diff
@@ -0,0 +1,98 @@
+limitations under the License.
+*/
+
+package shared
```
nit: i don't think we specifically need this in shared, it could just be a `tags` packageinstead

richardcase replied:
But lets leav it here for the time being. So no action needed on this comment.

salasberryfin replied:
My reason to put this in `shared` is that resource manager tags are used for different types of resources (compute, disks, clusters, etc.) and I didn't think of tags as resources at the level of other `cloud/services` like compute or container. Hopefully this makes sense.

richardcase replied:
Good point, thanks @salasberryfin

richardcase commented on exp/api/v1beta1/gcpmanagedcluster_types.go:
```diff
@@ -49,6 +49,12 @@ type GCPManagedClusterSpec struct {
 	// +optional
 	AdditionalLabels infrav1.Labels `json:"additionalLabels,omitempty"`
 
+	// resourceManagerTags is an optional set of tags to apply to GCP resources managed
```
```suggestion
	// ResourceManagerTags is an optional set of tags to apply to GCP resources managed
```

richardcase commented:
This is only adding optional API fields so:

/override pull-cluster-api-provider-gcp-apidiff

salasberryfin commented:
Thanks @richardcase. I'll be happy to work on #1053 as well.

richardcase commented:
Like before:

/override pull-cluster-api-provider-gcp-apidiff

richardcase commented:
Thanks @salasberryfin 

/lgtm
/approve
